flink-optimizer/src/main/java/org/apache/flink/optimizer/traversals/RangePartitionRewriter.java

```diff
@@ -0,0 +1,188 @@
+import org.apache.flink.api.common.typeutils.TypeComparatorFactory;
+import org.apache.flink.api.java.functions.AssignRangeIndex;
+import org.apache.flink.api.java.functions.PartitionIDRemoveWrapper;
+import org.apache.flink.api.java.functions.RangeBoundaryBuilder;
+import org.apache.flink.api.java.functions.SampleInCoordinator;
+import org.apache.flink.api.java.functions.SampleInPartition;
+import org.apache.flink.api.java.sampling.IntermediateSampleData;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.api.java.typeutils.TupleTypeInfo;
+import org.apache.flink.api.java.typeutils.TypeExtractor;
+import org.apache.flink.optimizer.dag.GroupReduceNode;
+import org.apache.flink.optimizer.dag.MapNode;
+import org.apache.flink.optimizer.dag.MapPartitionNode;
+import org.apache.flink.optimizer.dag.TempMode;
+import org.apache.flink.optimizer.plan.Channel;
+import org.apache.flink.optimizer.plan.NamedChannel;
+import org.apache.flink.optimizer.plan.OptimizedPlan;
+import org.apache.flink.optimizer.plan.PlanNode;
+import org.apache.flink.optimizer.plan.SingleInputPlanNode;
+import org.apache.flink.optimizer.util.Utils;
+import org.apache.flink.runtime.operators.DriverStrategy;
+import org.apache.flink.runtime.operators.shipping.ShipStrategyType;
+import org.apache.flink.util.Visitor;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+
+public class RangePartitionRewriter implements Visitor<PlanNode> {
+
+	OptimizedPlan plan;
+
+	public RangePartitionRewriter(OptimizedPlan plan) {
+		this.plan = plan;
+	}
+
+	@Override
+	public boolean preVisit(PlanNode visitable) {
+		return true;
+	}
+
+	@Override
+	public void postVisit(PlanNode visitable) {
+		List<Channel> outgoingChannels = visitable.getOutgoingChannels();
+		List<Channel> appendOutgoingChannels = new LinkedList<>();
+		List<Channel> removeOutgoingChannels = new LinkedList<>();
+		for (Channel channel : outgoingChannels) {
+			ShipStrategyType shipStrategy = channel.getShipStrategy();
+			if (shipStrategy == ShipStrategyType.PARTITION_RANGE) {
+				if (channel.getDataDistribution() == null) {
+					removeOutgoingChannels.add(channel);
+					appendOutgoingChannels.addAll(rewriteRangePartitionChannel(channel));
+				}
+			}
+		}
+		outgoingChannels.addAll(appendOutgoingChannels);
+		for (Channel channel : removeOutgoingChannels) {
+			outgoingChannels.remove(channel);
+		}
+	}
+
+	private List<Channel> rewriteRangePartitionChannel(Channel channel) {
+		List<Channel> appendOutgoingChannels = new LinkedList<>();
+		PlanNode sourceNode = channel.getSource();
+		PlanNode targetNode = channel.getTarget();
+		int sourceParallelism = sourceNode.getParallelism();
+		int targetParallelism = targetNode.getParallelism();
+		TypeComparatorFactory<?> comparator = Utils.getShipComparator(channel, this.plan.getOriginalPlan().getExecutionConfig());
+		// 1. Fixed size sample in each partitions.
+		long seed = org.apache.flink.api.java.Utils.RNG.nextLong();
```

Comment:
I would use deterministic value here. This will make the behavior more deterministic and testing the feature easier.